DRAFT feat(reset-password): Use OTP instead of link to start reset (front-end)

vpomerleau opened the pull request:
Putting up as draft for early review - functionality is in place and ready to test but need to add unit tests and functional tests

In this branch, reset password with code is enabled by default for dev via feature flags in auth server and content-server.

To note, this PR also includes applying container/presentation architectural pattern to the reset password pages.

Because:

*

This commit:

*

Closes #FXA-7890

## Because

-

## This pull request

-

## Issue that this pull request solves

Closes: # (issue number)

## Checklist

_Put an `x` in the boxes that apply_

- [ ] My commit is GPG signed.
- [ ] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

LZoog commented on packages/fxa-settings/src/components/App/index.tsx:
```diff
@@ -76,6 +76,10 @@ import PrimaryEmailVerified from '../../pages/Signup/PrimaryEmailVerified';
 import WebChannelExample from '../../pages/WebChannelExample';
 import LinkValidator from '../LinkValidator';
+import ResetPasswordStartContainer from '../../pages/ResetPasswordRedesign/ResetPasswordStart/container';
+import ConfirmResetPasswordOtpContainer from '../../pages/ResetPasswordRedesign/ConfirmResetPasswordOtp/container';
```
Since these are already in a new directory `ResetPasswordRedesign` that I assume we'll get rid of after we're stable, should we follow our convention of the directory name and component names roughly matching our route names, and then just rename the import here if needed? Having `Otp` especially in the names feels funky to me.

LZoog commented on packages/fxa-settings/src/components/LinkRememberPassword/index.tsx:
```diff
@@ -5,31 +5,33 @@
     params.set('email', email);
+    linkHref = `/signin?${params}`;
+  } else {
+    linkHref = params.size > 0 ? `/?${params}` : '/';
```
TIL `URLSearchParams().size`!

Out of scope here to update all the instances but we have to do this so much I feel like we should make a helper to do this for us so we don't have to think about it, we're doing `queryParams.length > 1 && queryParams) || ''` in at least one other spot...

LZoog commented on packages/fxa-settings/src/components/LinkRememberPassword/index.tsx:
```diff
@@ -5,31 +5,33 @@
+        <p>Remember your password?</p>
+      </FtlMsg>
+      <FtlMsg id="remember-password-signin-link">
+        {/* TODO replace with Link once React Signin fully rolled out */}
```
Hm, we should probably add a note about this to the 100% prod rollout ticket since it's easy enough.

LZoog commented on packages/fxa-settings/src/components/images/index.tsx:
```diff
@@ -143,3 +144,12 @@ export const LightbulbImage = ({ className, ariaHidden }: ImageProps) => (
+
+export const EmailCodeImage = ({ className, ariaHidden }: ImageProps) => (
+  <PreparedImage
+    ariaLabel="Illustration to represent an email containaing a code."
```
~containaing~ containing

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/AccountRecoveryConfirmKeyOtp/container.tsx:
```diff
@@ -0,0 +1,139 @@
+}: AccountRecoveryConfirmKeyOtpContainerProps & RouteComponentProps) => {
+  const [errorMessage, setErrorMessage] = useState('');
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const account = useAccount();
```
Can we avoid using `useAccount`? When used, it tries to fetch all of the account data that we use for Settings. OK in a follow up if that blows things up.

vpomerleau replied:
I've filed a follow-up: `FXA-9672`

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/AccountRecoveryConfirmKeyOtp/container.tsx:
```diff
@@ -0,0 +1,139 @@
+    estimatedSyncDeviceCount,
+    token,
+    uid,
+  } = location.state as AccountRecoveryKeyOtpLocationState;
```
We probably want an `|| {}` here similar to signin and signup to avoid breaking things if we try to destructure on an `undefined` (which shouldn't happen, but we can't guarantee).

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/AccountRecoveryConfirmKeyOtp/container.tsx:
```diff
@@ -0,0 +1,139 @@
+        },
+      });
+    } catch (error) {
+      // TODO in FXA-7890 clearer error handling
```
TODO in this PR?

vpomerleau replied:
I've cleared all of the straggler TODOs marked with this issue number. Thank you!

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/AccountRecoveryConfirmKeyOtp/index.tsx:
```diff
@@ -0,0 +1,167 @@
+        warningMessageFtlId="account-recovery-confirm-key-warning-message"
+        warningType="Note:"
+      >
+        If you reset your password and don't have account recovery key saved,
```
Probably WIP but some of these don't have FTL IDs. This is also a straight quote instead of curly.

vpomerleau replied:
The ftl id is passed in to the component here. The string already existed in the previous design, and I opted not to recreate new FTL entries since the text here is unchanged.

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPasswordOtp/container.tsx:
```diff
@@ -0,0 +1,207 @@
+
+  const handleNavigationWithRecoveryKey = () => {
+    navigate(`/reset_password_with_recovery_key_verified${location.search}`, {
+      replace: true,
```
Hm, I see where we're using `replace: true` in other places but we should only use it when we don't want the user to be able to navigate back to this page manually. Just checking that's the case here, maybe it is, but maybe for `handleNavigationWithoutRecoveryKey` we want them to be able to click back if they want to?

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPasswordOtp/container.tsx:
```diff
@@ -0,0 +1,207 @@
+    newPassword: string,
+    token: string
+  ) => {
+    // TODO move out of account model?
```
👍

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPasswordOtp/container.tsx:
```diff
@@ -0,0 +1,207 @@
+      // know what to hash the new password with. This is important in the scenario where a user
+      // has changed their primary email address. In this case, they must still hash with the
+      // account's original email because this will maintain backwards compatibility with
+      // how account password hashing works previously.
```
Probably worth noting this is only the case for the v1 key stretch, we're using a random salt for v2.

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPasswordOtp/container.tsx:
```diff
@@ -0,0 +1,207 @@
+  const notifyBrowserOfSignin = async (accountResetData: AccountResetData) => {
+    if (
+      isSyncDesktopV3Integration(integration) ||
+      isSyncOAuthIntegration(integration)
```
Instead of `isSyncDesktopV3Integration(integration) || isSyncOAuthIntegration(integration)`, we should use `integration.isSync()`. The OAuth integration is not always sync.

LZoog commented on packages/fxa-settings/src/pages/Signin/SigninTotpCode/index.tsx:
```diff
@@ -174,7 +174,7 @@ export const SigninTotpCode = ({
       <div className="mt-5 link-blue text-sm flex justify-between">
         <FtlMsg id="signin-totp-code-other-account-link">
-          <Link to={`/signin${location.search}`} className="text-start">
+          <Link to={`/${location.search}`} className="text-start">
```
If we need to change this to `/`, it's going to need to be a hard navigate. Though, if someone's got something in local storage they're going to be redirected back to `/signin`.

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/ResetPasswordStart/interfaces.ts:
```diff
@@ -0,0 +1,45 @@
+
+export interface ResetPasswordOAuthIntegration {
+  type: IntegrationType.OAuth;
+  getRedirectUri: () => ReturnType<OAuthIntegration['getRedirectUri']>;
```
We probably want to use `Pick` instead here, see the Signin interfaces like `type SigninOAuthIntegration`. I started doing it the way you have it and then Dan used `Pick` and I was like why didn't I do that originally, so I blame myself for this pattern being around😂

vpomerleau replied:
Oh yeah, that's much nicer.

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/ResetPasswordStart/index.tsx:
```diff
@@ -0,0 +1,142 @@
+
+        <FtlMsg id="password-reset-submit-button">
+          <button
+            data-testid="reset-password-button"
```
Does this need this `data-testid`?

vpomerleau replied:
Looks like no!

LZoog commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/ResetPasswordStart/container.tsx:
```diff
@@ -0,0 +1,61 @@
+      metricsContext,
+    };
+    try {
+      // TODO in FXA-7890 directly use auth client or not?
```
I think for the PW redesign we opted to just use auth-client? I feel like if there's a state update or a value we need to reference elsewhere it's nice to use GQL so we don't have to manually add things to the apollo cache but not sure that's relevant for most reset PW stuff.

LZoog reviewed (commented):
Phew, there's a lot going on here, thanks for all your work. This is just my initial code review, I'll check into the branch and play around with it on the next pass through.

vpomerleau commented on packages/fxa-auth-server/config/dev.json:
```diff
@@ -467,5 +467,8 @@
     "firstInterval": "5s",
     "secondInterval": "10s"
+  },
+  "passwordForgotOtp": {
```
Good question, I was wondering if we could have just one environment variable on the SRE side that is used for both auth and content flags. We definitely want to have both on or off at the same time.

vpomerleau replied:
I think I'll leave these separate and keep the passwordForgotOtp name because it's closely tied to all the name choices for the auth endpoint.

dschom commented on packages/functional-tests/tests/misc/keyStretchingV2.spec.ts:
```diff
@@ -14,6 +14,7 @@ import { RecoveryKeyPage } from '../../pages/settings/recoveryKey';
 import { SignupReactPage } from '../../pages/signupReact';
 import { TotpPage } from '../../pages/settings/totp';
 import { getCode } from 'fxa-settings/src/lib/totp';
+import { ConfigPage } from '../../pages/config';
```
These changes are going to conflict with, https://github.com/mozilla/fxa/pull/16923. Might want to coordinate with Kat to see which changes make the most sense to land first.

dschom commented on packages/functional-tests/tests/misc/keyStretchingV2.spec.ts:
```diff
@@ -532,7 +576,14 @@ test.describe('key-stretching-v2', () => {
       page,
       target,
-      pages: { signupReact, settings, login, recoveryKey, resetPasswordReact },
+      pages: {
```
nit: stray new line?

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/AccountRecoveryConfirmKey/index.test.tsx:
```diff
@@ -0,0 +1,195 @@
+import { MOCK_RECOVERY_KEY } from '../../mocks';
+import GleanMetrics from '../../../lib/glean';
+
+const mockVerifyRecoveryKey = jest.fn((recoveryKey: string) =>
```
Warning about recoveryKey is declared but never used. You can remove it, or prefix with _ to get rid of warning.

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/AccountRecoveryConfirmKey/index.tsx:
```diff
@@ -0,0 +1,167 @@
+  setIsSubmitting,
+  verifyRecoveryKey,
+  token,
+  uid,
```
Never used

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/AccountRecoveryConfirmKey/interfaces.ts:
```diff
@@ -0,0 +1,33 @@
+  serviceName: MozServices;
+  setErrorMessage: React.Dispatch<React.SetStateAction<string>>;
+  setIsSubmitting: React.Dispatch<React.SetStateAction<boolean>>;
+  verifyRecoveryKey: (recoverkyKey: string) => Promise<void>;
```
typo recoverkyKey > recoveryKey

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPassword/container.tsx:
```diff
@@ -0,0 +1,195 @@
+  const isResetWithoutRecoveryKey = !!(code && token);
+
+  const handleNavigationWithRecoveryKey = () => {
+    navigate(`/reset_password_with_recovery_key_verified${location.search}`);
```
Just curious, in other places you reference `searchParams` (which come from `location.search`), but here you reference `location.search`. Is there a reason for not using searchParams here?

vpomerleau replied:
No reason, just missed the update.

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPassword/index.test.tsx:
```diff
@@ -0,0 +1,156 @@
+      expect(link).toBeVisible();
+      expect(link).toHaveAttribute(
+        'href',
+        '/signin?email=johndope%40example.com'
```
maybe use encodeURIComponent(MOCK_EMAIL) instead?

vpomerleau replied:
💯

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/ResetPassword/container.tsx:
```diff
@@ -0,0 +1,59 @@
+        state: { email, metricsContext },
+      });
+    } catch (err) {
+      localizedErrorMessage = getLocalizedErrorMessage(ftlMsgResolver, err);
```
Do we expect this to happen? Perhaps record Sentry event if this is unexpected so we can track it?

vpomerleau replied:
This could happen due to throttling on `passwordForgotSendOtp`, for example, so we would expect to show a user-facing message in that instance. If there are issues on the backend, those would already be recorded in Sentry, correct?

dschom replied:
Any unexpected error would get recorded by `auth-server`; however, expected error states like throttling errors (from customs), or even a payload validation error that you might not have expected to occur would not. It sounds like we have good handle on the potential failures here and it's not necessary.

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPassword/index.test.tsx:
```diff
@@ -0,0 +1,156 @@
+import { renderWithLocalizationProvider } from 'fxa-react/lib/test-utils/localizationProvider';
```
Missing license comment block.

dschom commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/ConfirmResetPassword/index.tsx:
```diff
@@ -0,0 +1,93 @@
+  return (
+    <AppLayout>
+      <FtlMsg id="confirm-reset-password-otp-flow-name">
+        <p className="text-start text-grey-400 text-sm">Reset password</p>
```
This looks off. Just double checking that the CSS is as intended here.

vpomerleau replied:
This matches the Figma design. I went back and forth on marking it as a heading considering the small font.

![image](https://github.com/mozilla/fxa/assets/22231637/670d416f-1b14-4200-8d87-098160dab40c)

dschom replied:
Sounds good. Just double checking!

vpomerleau commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPassword/index.test.tsx:
```diff
@@ -0,0 +1,156 @@
+      );
+      expect(
+        screen.getByText(
+          'When you reset your password, you reset your account. You may lose some of your personal information (including history, bookmarks, and passwords). That’s because we encrypt your data with your password to protect your privacy. You’ll still keep any subscriptions you may have and Pocket data will not be affected.'
```
Agree it's not ideal. These error message will be removed/replaced in `FXA-7901` though with components that will provide better options for accessible locators. Maybe okay to keep as is for now?

vpomerleau commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/ConfirmResetPassword/container.tsx:
```diff
@@ -0,0 +1,145 @@
+    const options = { metricsContext };
+    try {
+      const { code, emailToHashWith, token, uid } =
+        await authClient.passwordForgotVerifyOtp(email, otpCode, options);
```
I was wondering about that too. @LZoog replied here that we might be okay in this instance to just directly use auth-client for reset password: https://github.com/mozilla/fxa/pull/16869#discussion_r1596066851

vpomerleau commented on packages/fxa-settings/src/pages/ResetPasswordRedesign/ConfirmResetPassword/index.tsx:
```diff
@@ -0,0 +1,93 @@
+        elems={{ span: spanElement }}
+      >
+        <p className="text-start">
+          Enter the 8-digit confirmation code we sent to {spanElement}
```
I've asked Eduardo if we should add a note about the code duration here. Will update based on his recommendation!

vpomerleau commented:
Thanks very much for your review @dschom ! ❤️ I think I've addresses all comments - just waiting on CI tests and coordinating with Kat about the key stretching functional tests. I went through the flow with account recovery key and no longer seeing the warning. If you have a chance to check back, would be very appreciated!

vpomerleau commented on packages/fxa-auth-server/test/mail_helper.js:
```diff
@@ -62,6 +63,8 @@ module.exports = (printLogs) => {
             console.log('\x1B[36mUnblock code:', uc, '\x1B[39m');
             console.log('\x1B[36mReport link:', rul, '\x1B[39m');
+          } else if (rpc) {
+            console.log('\x1B[36mReset password Otp:', rpc, '\x1B[39m');
```
@chenba similar to the mail helper you have in a separate PR. Should I remove from here and r+ yours?
